File: src/main/kotlin/com/store/model/Product.java
package com.store.model;

public class Product {
    public final int id;
    public final String name;
    public final ProductType type;
    public final int inventory;
    public final double cost;

    public Product(ProductId productId, ProductDetails details) {
        this.id = productId.id;
        this.name = details.name;
        this.type = details.type;
        this.inventory = details.inventory;
        this.cost = details.cost;
    }
}

File: src/main/kotlin/com/store/model/Product.kt
package com.store.model

import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import java.util.concurrent.atomic.AtomicInteger
import javax.validation.constraints.NotNull
import javax.validation.constraints.Positive

data class Product(
    @field:NotNull @field:JsonDeserialize(using = StrictStringDeserializer::class) val name: String = "",
    @field:NotNull val type: String = "gadget",
    @field:NotNull @field:Positive val inventory: Int = 0,
    @field:NotNull @field:Positive val cost: Int = 0,
    val id: Int = idGenerator.getAndIncrement()
) {
    companion object {
        val idGenerator: AtomicInteger = AtomicInteger()
    }
}

enum class ProductType {
    book,
    food,
    gadget,
    other
}